decrypt.go

```diff
@@ -31,8 +33,31 @@ func (p7 *PKCS7) Decrypt(cert *x509.Certificate, pkey crypto.PrivateKey) ([]byte
 	}
 	switch pkey := pkey.(type) {
 	case crypto.Decrypter:
-		// Generic case to handle anything that provides the crypto.Decrypter interface.
-		contentKey, err := pkey.Decrypt(rand.Reader, recipient.EncryptedKey, nil)
+		var opts crypto.DecrypterOpts = nil
+		switch algorithm := recipient.KeyEncryptionAlgorithm.Algorithm; {
+		case algorithm.Equal(OIDEncryptionAlgorithmRSAESOAEP):
+			var hashFunc crypto.Hash
+			hashFunc, err := getHashFuncForKeyEncryptionAlgorithm(recipient.KeyEncryptionAlgorithm)
+			if err != nil {
+				return nil, err
+			}
+			opts = &rsa.OAEPOptions{Hash: hashFunc}
+		case algorithm.Equal(OIDEncryptionAlgorithmRSAMD5):
+			opts = &rsa.OAEPOptions{Hash: crypto.MD5}
+		case algorithm.Equal(OIDEncryptionAlgorithmRSASHA1):
+			opts = &rsa.OAEPOptions{Hash: crypto.SHA1}
+		case algorithm.Equal(OIDEncryptionAlgorithmRSASHA256):
+			opts = &rsa.OAEPOptions{Hash: crypto.SHA256}
+		case algorithm.Equal(OIDEncryptionAlgorithmRSASHA384):
+			opts = &rsa.OAEPOptions{Hash: crypto.SHA384}
+		case algorithm.Equal(OIDEncryptionAlgorithmRSASHA512):
+			opts = &rsa.OAEPOptions{Hash: crypto.SHA512}
```

Comment:
SHA224, 1.2.840.113549.1.1.14.

Reply:
Added in 76fd9de0905ef97ce7dd9b1e9156c0bd550c14d1